Mvp kiss

hellais opened the pull request:
I think it's more reasonable to roll out the golang based collector with an incremental approach:

1. Write a MVP Keep It Simple Silly® version of a golang collector that is identical to our current python one
2. Write good test coverage of it
3. Ensure we have good logging to detect defects (with sentry)
4. Deploy it as a replacement for the current collector on one of our two collectors
5. If all is good deprecate the legacy python based collector

Once we are confident that this doesn't create any issues and are OK with this codebase we can build upon this smaller neater golang codebase.

As part of this I written this collector in ~800 loc including tests and it only supports the most basic functionality of the current collector.

It currently has 68% code coverage, but my aim is to reach ~90% of unittest coverage and have some solid integration tests which start measurement_kit and probe-legacy to ensure that measurements are collected properly.

hellais committed:
Flatten the package hierarchy

hellais committed:
Minimalistic collector implementation which works just like the current one

hellais committed:
Start writing unittests for collector

hellais committed:
Reach 68% code coverage

hellais committed:
Add better support for testing time base functions

* Drop currently untested/unused single measurement submission API

hellais committed:
Bump dependencies

hellais committed:
Refactor the unittests

hellais committed:
Write tests for the expiration of reports

hellais committed:
Add more tests

hellais commented:
I am pretty happy with the level of unittest coverage I have now and it was useful to spot a couple more bugs.

Currently I don't support the legacy yaml based format, but we could possibly opt for dropping it.
@bassosimone @darkk what do you think of dropping YAML format support? Do you recall what the conclusion of that discussion was?

My next step for this is writing some integration tests that run measurement_kit and probe-legacy in a docker VM to check if they work as expected.

It would be useful in the meantime to get some code review of this branch.

It's probably most agile to simply clone the repo and read the code directly (rather than looking at the diffs in here).

bassosimone commented on collector/api.go:
```diff
@@ -0,0 +1,41 @@
+	v1.POST("/report/:reportID", UpdateReportHandler)
+	v1.POST("/report/:reportID/close", CloseReportHandler)
+	// XXX not implemented
+	// v1.POST("/measurement", SubmitMeasurementHandler)
```
What is this supposed to do?

hellais replied:
This is supposed to be the single request, single measurement submit handler. See: https://github.com/ooni/spec/blob/master/backends/bk-001-ooni-backend.md#2314-submitting-single-measurements-using-a-single-api-call.

I commented it out for the time being for the sake of keeping this an MVP which is just an implementation of the current collector.

bassosimone replied:
Could you use actionable comments like `TODO(hellais): This is $thing and is currently commented out because $reason`?

bassosimone replied:
I have changed this into an actionable comment. I've added it as a TODO for myself since that's something I will most likely end up doing.

bassosimone commented on collector/collector_test.go:
```diff
@@ -0,0 +1,324 @@
+			"platform": "android",
+		},
+		DataFormatVersion: "0.2.0",
+		Input:             "http://google.com",
```
My understanding is that the data format is such that `input` is either a non empty string or `null`. Is the new collector going to enforce this different constraing that `input` is a string?

hellais replied:
This is a very good point. As it is now the `input` key is defined as a string, so we do not support it being `null`. We should add a test case for a `null` `input` and handle it appropriately.

bassosimone replied:
I am sad that our legacy forces us to write convoluted code. Please, write a test case for this specific scenario. Also, may I suggest to change the spec to remove, in going forward, all the possible type related ambiguity? May I suggest to bless a new data format version that just clarifies what the expected type would be? I will be super happy to make changes in MK to make this happen, as the possibility of having different types is also creating pain for me.

(Not sure whether this comment makes sense, but IIRC there are also Golang libraries that allow for more flexible parsing of JSON. If it becomes too hairy to deal with several formats, we should probably consider these. TBH, my preferred approach would be to use the data format version to select whether to do type-clean processing or the less-type-clean processing using a more flexible parser. I'd also like to have maybe some way of measuring the amount of old measurements that are submitted, so we decide when to pull the plug. This does not necessarily need to happen here, but I'd say we should have a plan for moving forward faster.)

bassosimone replied:
So, it turns out that we don't need to concern ourselves about `null` values in input:

1. `go` is nice enough that it ignores `null`s in input

2. as long as we store the blog we received and we only use the go code for parsing

This seems quite neat, as it requires less code.

bassosimone commented on collector/handler.go:
```diff
@@ -0,0 +1,139 @@
+		req.Format = supportedFormats[0]
+	}
+	if stringInSlice(req.Format, supportedFormats) != true {
+		c.JSON(http.StatusNotAcceptable, gin.H{"error": "invalid format. Must be either json or yaml"})
```
Are we still supporting YAML? (Do we still have YAML users?)

hellais replied:
In this PR I did not keep support for YAML. Before rolling this out we should evaluate if we indeed want to keep supporting YAML (it's probably not too much effort to add YAML support) or if we are OK with dropping it.

I just looked up the stats for submitted measurements in the month of January 2019 (JSON reporting was added in ooniprobe 1.4.0):
```
3	ooniprobe	2.2.0.rc2
85	ooniprobe	2.3.0-rc.1
420	ooniprobe	2.2.1-dev
21093	ooniprobe	2.2.0
105441	ooniprobe	2.1.0
337	ooniprobe	2.0.2
144	ooniprobe	1.6.1
156	ooniprobe	1.5.1
336	ooniprobe	1.3.2
```

This is a pretty small fraction of the total OONI Probe desktop users and given that measurement from such an old version of ooniprobe (for ex web_connectivity did not exist then) are also not that useful I am airing on the side of dropping support for it.

bassosimone replied:
I am going to resolve this. As mentioned, we won't implement YAML: https://github.com/ooni/collector/pull/3#issuecomment-501757442. The above comment explains why we won't bother doing that.

bassosimone commented on collector/handler.go:
```diff
@@ -0,0 +1,139 @@
+	reportID, err := CreateNewReport(store, req.Format)
+	if err != nil {
+		// XXX check this against the spec
+		c.JSON(http.StatusNotAcceptable, gin.H{
```
Is there any client relying on the returned value? I always coded all clients such that if the return value is not `200` then the whole operation has failed and we are sad and try to move on. (I understand that once the spec is written it is written, but if we know that there are no clients that we don't control and/or rely on this behaviour, I would just relax the spec to say that it returns either `StatusOK` or `StatusNotAcceptable`.)

hellais replied:
Yeah I think that probably relaxing the spec on this is a good idea. Anything other than a 2xx message should always be considered as a failure by a well written client.

bassosimone replied:
Yes, please. I'd suggest to open a PR for the spec. I'll review it within an hour.

bassosimone replied:
OTOH, it's probably a good practice to return a valid JSON.

bassosimone commented on collector/handler.go:
```diff
@@ -0,0 +1,139 @@
+
+// DeprecatedUpdateReportHandler is for legacy clients
+func DeprecatedUpdateReportHandler(c *gin.Context) {
+	c.JSON(http.StatusOK, gin.H{
```
Is it intended that we basically discard the request? Are there any plans to implement this code? If there are no plans, what is the reason to pretend we're accepting legacy client submissions?

hellais replied:
Thanks for pointing this out. I need to double check this, I suspect it may be that since we don't support YAML there is no point in having this here, but if that is the case I agree we should drop this code and not pretend it works.

bassosimone replied:
As discussed in https://github.com/ooni/collector/pull/3#issuecomment-501759144, we'll drop this code.

bassosimone commented on collector/handler.go:
```diff
@@ -0,0 +1,139 @@
+	if err != nil {
+		if err == ErrReportNotFound {
+			log.WithError(err).Debug("report not found error")
+			// XXX use the correct return value
```
Same comment as above wrt the return value

bassosimone commented on collector/handler.go:
```diff
@@ -0,0 +1,139 @@
+	}
+	// XXX temporarily disabled
+	//platformMetric.MetricCollector.(*prometheus.CounterVec).WithLabelValues(meta.Platform).Inc()
+	//countryMetric.MetricCollector.(*prometheus.CounterVec).WithLabelValues(meta.ProbeCC).Inc()
```
Why temporarily disabled? Do we need to enable it for merging? Do we need to keep it around commented out.

hellais replied:
Thanks for pointing this out. I don't recall why it was commented it out, but I think we should re-enable it.

These are metrics that can be scraped by our monitoring and should aid us in investigating issues with measurement counts dropping suddenly from a particular country or platform version.

bassosimone replied:
Okay, then, please re-enable them

bassosimone commented on collector/handler.go:
```diff
@@ -0,0 +1,139 @@
+
+	err := CloseReport(store, reportID)
+	if err != nil {
+		// XXX return proper error
```
Same comment as above.

bassosimone commented on collector/report.go:
```diff
@@ -0,0 +1,331 @@
+	}
+}
+
+// ActiveReport stores metadata about an active report. The metadat is immutable
```
`s/metadat/metadata/`

bassosimone commented on collector/report.go:
```diff
@@ -0,0 +1,331 @@
+func GenReportID() string {
+	return fmt.Sprintf("%s_%s_%s",
+		time.Now().UTC().Format(TimestampFormat),
+		"AS00",
```
I am confused by this `AS00`.

hellais replied:
I didn't want to make knowing the AS of the probe at this point in the code a requirement (as it complicates the code), but to preserve compatibility with the current `report_id` I decided to put in the magic number `AS00`, which will make it so it's a mostly valid AS format, but at the same time possibly filterable to look for report_ids from new collectors in the DB.

I should probably add a comment in the code mentioning this to make it clear why it's there though.

bassosimone replied:
Yeah, please clarify. I am not even 100% sure I understood your above point.

bassosimone commented on collector/report.go:
```diff
@@ -0,0 +1,331 @@
+	SoftwareVersion      string       `json:"software_version"`
+	TestHelpers          interface{}  `json:"test_helpers"`
+	TestKeys             interface{}  `json:"test_keys"`
+	TestRuntime          float64      `json:"test_runtime"`
```
The above is potentially a breaking change. We did not enforce strict typing until now. Can you please update the specification to make sure that the type of each field is clearly specified? Have you tried to generate reports with old software (or just resubmit old reports) to make sure that it is still working properly? Can you write an integration test where sample reports submitted by old software are resubmitted and we make sure they are accepted? My main concern is that `input` is `null` for tests without input.

hellais replied:
> Can you write an integration test where sample reports submitted by old software are resubmitted and we make sure they are accepted? My main concern is that input is null for tests without input

This is a very good point and it's something that I had on my mind as something to try out. Perhaps a similar docker based testing strategy should be adopter here too and possibly run various OONI Probe version to validate that everything works as expected.

bassosimone replied:
I believe that is the super pro way of doing stuff. At the same time, I think it's also probably a bit too costly for you. A reasonable middle ground could be to have a `testdata` directory with reports that are originating from old OONI versions and then make sure they're accepted.

bassosimone commented on collector/report.go:
```diff
@@ -0,0 +1,331 @@
+	TestName             string       `json:"test_name"`
+	TestVersion          string       `json:"test_version"`
+	MeasurementStartTime string       `json:"measurement_start_time,omitempty"`
+	TestStartTime        string       `json:"test_start_time"` // XXX these should actually be time
```
What is the implication of this `XXX`?

hellais replied:
Perhaps it's actually a good idea to keep this as string as using `time.Time` will mean that we are parsing and re-serializing the test_start_time value and this could lead to inconsistent results.

In general we should probably do less transformation on the submitted data to avoid introducing minor bugs or issues in the resulting submitted data (by parsing and reserializing it).

bassosimone replied:
It also depends on whether you want to enforce the data having a specific format or not. If you do plan on doing that, but at the same time you don't want to introduce processing errors, the right approach sounds like parsing the string for correctness but then discarding the parsed date and from then on just continue using the validated string.

bassosimone commented on collector/report.go:
```diff
@@ -0,0 +1,331 @@
+	TestVersion          string       `json:"test_version"`
+	MeasurementStartTime string       `json:"measurement_start_time,omitempty"`
+	TestStartTime        string       `json:"test_start_time"` // XXX these should actually be time
+	Annotations          interface{}  `json:"annotations"`
```
Annotations are `map[string]string` in MK. What are they in OONI legacy?

hellais replied:
I believe we never really enforced anything as strict as that in OONI legacy, hence we are using the opaque `interface{}` in here for it.

bassosimone replied:
How about `map[string]interface{}` then?

bassosimone commented on collector/report.go:
```diff
@@ -0,0 +1,331 @@
+	Options              []string     `json:"options"`
+	ProbeASN             string       `json:"probe_asn"`
+	ProbeCC              string       `json:"probe_cc"`
+	ProbeCity            string       `json:"probe_city"`
```
Also, `ProbeCity` may be `null` (and MK always sets it to `null`).

hellais replied:
Good point. Probably a similar strategy to `input` needs to be applied to this as well.

bassosimone replied:
Right. I am wondering whether we care deeply about this field. Also, I think we can probably have a bump in the data version where we define this as strictly a string.

bassosimone commented on collector/report.go:
```diff
@@ -0,0 +1,331 @@
+	ProbeIP              string       `json:"probe_ip"`
+	SoftwareName         string       `json:"software_name"`
+	SoftwareVersion      string       `json:"software_version"`
+	TestHelpers          interface{}  `json:"test_helpers"`
```
This may probably be `map[string]interface{}` right?

hellais replied:
That could be, though I would rather we keep these more complex fields as opaque `interface{}` to avoid triggering possible parsing errors.

Maybe we should add an extra check in the code (or maybe do some work on the pipeline to understand how many measurement if any at all don't follow this definition).

bassosimone replied:
I would like us to start enforcing more data correctness at the collector level. I think a sound strategy here is to have data points coming from OONI versions that we care about the most and then make sure we can parse them.

bassosimone commented on collector/report.go:
```diff
@@ -0,0 +1,331 @@
+// CreateNewReport creates a new report
+func CreateNewReport(store *Storage, format string) (string, error) {
+	activeReport := NewActiveReport(format)
+	// XXX put the report metadata into an activeReports in memory map
```
What does this `XXX` mean?

hellais replied:
I believe this is no longer relevant (it is added to the `activeReport` map below), but I will double check that.

bassosimone replied:
Thanks. Please remove the comment if it's outdated.

bassosimone reviewed (changes requested):
After reading the diff, I have concerns regarding the data format. Please, address my concerns.

Also, please, resubmit 79ec2b62936a9806fcefff24bf35e33bc26283f4 as another PR. I would like to quickly review that and have it merged so we can reason on the more complex diff.

hellais commented:
@bassosimone thanks a bunch for your very thorough review, it's greatly appreciated!

I believe I have addressed all your comments and here is a checklist for what I plan to do to follow up on them:

* [ ] Come up with a way of handling nullable values in the submission and write tests for them (ex. Input, probe_city)
* [x] Decide on what we should do in relation to YAML and document the decision
* [ ] Update OONI Spec in relation to error conditions
* [x] Check what should be done about DeprecatedUpdateReportHandler
* [ ] Enable the prometheus metrics
* [ ] Fix minor typos in doc strings
* [ ] Improve docs for AS00
* [ ] Write end to end integration tests that use legacy version of OONI Probe and validate if the reports are submitted properly
* [ ] Review the MeasurementEntry schema and see if we can make it more robust
* [ ] Ensure all the XXX are resolved and if so update the docstring

bassosimone commented:
@hellais I think the main convo we need to have is related to an incremental strategy for increasingly enforcing more type correctness in what comes in input. We should probably have a quick chat.

bassosimone commented:
So, recently @hellais suggested that I should probably take responsibility here and make sure this is merged, which is also incidentally aligned with my plan at https://github.com/ooni/probe/issues/863.

Let's start this process by stating that with respect to

> Decide on what we should do in relation to YAML and document the decision

we WILL NOT implement YAML here.

bassosimone commented:
Here's another thing do decide:

>  Check what should be done about DeprecatedUpdateReportHandler

I am inclined to drop support for this functionality. (This requires a spec bump, but whatever, it basically boils down to saying in the spec that new clients MUST NOT use that feature.)

bassosimone committed:
Update api.go

bassosimone commented:
I have done a first pass at understanding what we have here. It seems to me this branch describes reasonably well the point where we want to arrive. I'll do another pass tomorrow.
